Source/CarthageKit/Xcode.swift

```diff
@@ -1225,26 +1285,50 @@ public func buildInDirectory(directoryURL: NSURL, withOptions options: BuildOpti
 					} else {
 						filteredSDKs = sdks.filter { options.platforms.contains($0.platform) }
 					}
-
 					return sdkFilter(sdks: filteredSDKs, scheme: scheme, configuration: configuration, project: project)
 				}
-
-				let buildProgress = buildScheme(scheme, withConfiguration: options.configuration, inProject: project, workingDirectoryURL: directoryURL, derivedDataPath: options.derivedDataPath, toolchain: options.toolchain, sdkFilter: wrappedSDKFilter)
-					// Discard any existing Success values, since we want to
+				
+				return buildScheme(scheme, withOptions: options, inProject: project, workingDirectoryURL: directoryURL, dependency: dependency, sdkFilter: wrappedSDKFilter)
+					.on(started: {
+						observer.sendNext(.Success(initialValue))
+					})
+			}
+			.startWithSignal({ (signal, signalDisposable) in
+				disposable += signalDisposable
+				
+				let ignoredValue = (ProjectLocator.Workspace(NSURL(string: "")!), "")
+				
+				let eventSignal: Signal<TaskEvent<(ProjectLocator, String)>, CarthageError> = signal
+					// Discard any Success values, since we want to
 					// use our initial value instead of waiting for
 					// completion.
-					.map { taskEvent in
-						return taskEvent.map { _ in initialValue }
+					.map { (taskEvent: TaskEvent<NSURL>) -> TaskEvent<(ProjectLocator, String)> in
+						return taskEvent.map { _ in ignoredValue }
 					}
-					.filter { taskEvent in taskEvent.value == nil }
-
-				return BuildSchemeProducer(value: .Success(initialValue))
-					.concat(buildProgress)
-			}
-			.startWithSignal { signal, signalDisposable in
-				disposable += signalDisposable
-				signal.observe(observer)
-			}
+					.filter { taskEvent in
+						return taskEvent.value == nil
+					}
+				
+				let versionFileSignal: Signal<TaskEvent<(ProjectLocator, String)>, CarthageError> = signal
+						.ignoreTaskData()
+						.collect()
+						.on(next: { (urls: [NSURL]) in
+							guard let dependency = dependency, rootDirectoryURL = rootDirectoryURL else {
+								return
+							}
+							if !createVersionFileForDependency(dependency, forPlatforms: options.platforms, buildProductURLs: urls, rootDirectoryURL: rootDirectoryURL) {
+								NSLog("Warning: Version file could not be created for \(dependency.project.name)")
+							}
+						})
+						.map { (urls: [NSURL]) -> TaskEvent<(ProjectLocator, String)> in
+							return .Success(ignoredValue)
+						}
+						.filter { taskEvent in
+							return false
+						}
+				
+				Signal.merge([eventSignal, versionFileSignal]).observe(observer)
```

Comment:
In that case, you should be able to use `flatMapTaskEvents`? (Or whatever that's called.)